standards/rails.md

```diff
@@ -0,0 +1,163 @@
+* Method parameters should be value objects, either to be operated on or needed as input
+* Methods should return rich result objects and not booleans
+* Dependent service objects should be accessible via private methods, and created either in the constructor or lazily
+
+#### Service Class Example
+
+```ruby
+# app/services/product_import_service.rb
+
+class ProductImportService
+
+  # Initialize the service with any necessary (or optional) information
+  def initialize(connection: nil)
+    @connection = connection || Connection.new
+  end
+
+  # Use a main public method
+  def process
+    # Do stuff
+  end
+
+  private
+  # Additional supporting methods belong here
+
+end
+```
+
+## Namespacing
+
+For larger projects it's often beneficial to namespace related models, controllers, and views into sub-folders to help with organization. There are no hard and fast rules but it's helpful to be familiar with how to do it.
+
+For an in-depth discussion on why namespacing is valuable and strategies on how to organize projects, check out [Makandra's blog](https://blog.makandra.com/2014/12/organizing-large-rails-projects-with-namespaces/).
+
+#### Namespacing Example
+```
+# app/models
+
+/activity
+/api
+/contact
+/invoice
+/planner
+/report
+/project
+activity.rb
+contact.rb
+planner.rb
+invoice.rb
+project.rb
+user.rb
+```
+
+## Models
+
+* Keep your models small
+* Logic that can be cleanly collected into a specific **business concern** should be split out into a **service class**
+  * e.g. `Product.import_products` should be refactored into `ProductImportService`
+* Use models for `has_and_belongs_to_many:` relationships, e.g. `has_many: :a_things, through: :a_b_things`
+* Namespace related models where it makes sense to do so
+
+## Controllers
+
+* Keep your controllers small - they control things, they don't **do** things
+* Use restful routing as much as possible. Avoid non-CRUD actions unless you have a compelling reason
+
+## Front End
+
+The below guidelines are for work done in a team of Rails developers, designed with maintainability and ease of project ramp-up for new Rails people in mind.
+
+When working with a front end developer, talk with them to see what standards and tooling they like to use.
+
+### CSS
+
+* Use SCSS for all your files, including `application.scss`
+* Don't use controller-generated asset files
+* Keep your CSS modular and component-based
+* Use SCSS variables for things like colors, generic padding, and spacing, and keep them in a `_vars.scss` file
+
+A typical folder setup will look something like this:
+
+```
+base/
+  _elements.scss
+  _fonts.scss
+  _reset.scss
+helpers/
+  _utils.scss
+  _vars.scss
+landmarks/
+  _header.scss
+  ...
+modules/
+  _btn.scss
+  ...
+_screen.scss # loads everything in subfolders
+application.scss # loads _screen.scss, and any other applicable stylesheets
+```
+
+### Javascript
+
+* Use Javascript, not Coffeescript
+* Use Yarn to load Javascript libraries in Rails 5
```

Comment:
Why yarn?  Since `npm@5.0.0`, there is very little difference between npm and yarn.  Npm comes with Node, so enforcing yarn effectively equates to mandating an additional project dependency.

It may be beneficial to also call out _how_ to install node.  Using Brew is not the best way, instead, a node version manager like [nvm](https://github.com/creationix/nvm) is the better option (for the same reasons rvm is a better way to install Ruby)

Reply:
Npm has improved, but, is still significantly slower than yarn according to the sources I've found.

That being said, I do believe we can recommend tools, but mandating them should be left for situations in which they improve the workflow significantly for the entire discipline.  Yarn vs npm probably doesn't make a significant difference in overall workflow, but, I welcome information telling me otherwise. :)

Just for my two cents, specifically in regards to npm vs yarn.

Reply:
For a while (in js timescale) yarn was the absolute winner; faster installs, a `.lock` file, etc.  But, that was sooo last year.  Today, the two are really pretty close: https://medium.com/netscape/npm-5-yarn-killer-ba69737b24d0.

My argument with Yarn vs Npm is threefold:
1) Yarn adds another dependency
1) It doesn't really provide much extra value over npm
1) If you know the difference between yarn and npm, you probably already know what you're doing and can make your own decisions.

Ultimately, I agree with you @powder, mandating tools belongs in the context of a project.